### PR TITLE
Cache territory name bounds.

### PR DESCRIPTION
**Pull request by asvitkine**
## Overview
Cache territory name bounds.

Profiling revealed this part is expensive, because to find the
best rectangle it tests whether a bunch of different rectangles
intersect with the territory bounds.

Looks like it can be easily cached however.

## Functional Changes

No user visible changes intended.

## Manual Testing Performed
- Load Feudal Japan map and check that map names render OK. Change zoom levels.
- Changing font sizes: On World at War map, try changing font size - for example from 12 to 20 and check that Kamchatka territory's name's location changes appropriately with the size change.

### CODE REVIEW

**Comment by ron-murhammer** [comment]
Did you try changing font sizes as well? That appears to influence the calculation and might want to consider clearing the cached value if the font size is changed.

**Comment by asvitkine** [comment]
Just tried it, and it seems to work fine. I actually don't know why. Maybe something re-creates this object when that happens?

A good territory to try this with is Kamchatka on World at War map, where at lower font sizes the name is further down, but at higher font sizes it's raised.

**Comment by asvitkine** [comment]
Actually, maybe this optimization doesn't work? Looks like in TileManager.drawTerritory(), this object gets re-created each time?

If so, let me take a look at doing it a different way...

**Comment by asvitkine** [comment]
Just checked, and TileManager.drawTerritory() doesn't get called each time we draw a territory - it only gets called once on game load for each territory (to set up the objects to draw it). And doesn't get called anymore until something changes like the font size.

So this patch works as expected and the optimization is indeed applied.

**Comment by ron-murhammer** [comment]
@asvitkine Yeah, it appears `MapPanel.resetMap()` clears all the drawables and is called during font change as well as a number of other changes to the map.

Did you do any kind of small benchmark to show this makes a performance improvement? That method is heavy and even showing reduced calls to it would be enough to verify. I'd just like some basic either number of calls or timing data to show this is worth adding.

**Comment by asvitkine** [comment]
Sure. I instrumented TerritoryNameDrawable.draw() to count number of execution times and total time spent in that function, making the game print stats and exit after 1000 executions.

On Feudal Japan with players set to AI, it went from the first 1000 calls taking a total of 1500ms to a total of 400ms on my machine.

**Comment by asvitkine** [comment]
Also savings are likely higher with a greater number of executions since probably a lot of the initial ones are uncached.

File: game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
 public class TerritoryNameDrawable implements IDrawable {
   private final String territoryName;
   private final UiContext uiContext;
+  private Rectangle territoryBounds;
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java]: You might want to wrap this in a SoftReference. 
This way it still can be garbage collected when the machine absolutely requires some memory.

**Comment by asvitkine** [comment]
The object is Rectangle which doesn't use much memory. I think
SoftReference is overkill here.

File: game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java [hunk 2]
```diff
@@ -80,7 +81,10 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
       x = namePlace.get().x;
       y = namePlace.get().y;
     } else {
-      final Rectangle territoryBounds = getBestTerritoryNameRect(mapData, territory, fm);
+      if (territoryBounds == null) {
+        // Cache the bounds since re-computing it is expensive.
+        territoryBounds = getBestTerritoryNameRect(mapData, territory, fm);
+      }
```
Comment (ssoloff) [inline game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java]: Via tracing, it appears that `draw()` can be called by multiple threads.  In that case, you might want to consider some form of synchronization when reading/writing `territoryBounds`.  Without it, there's the possibility that the cached reference won't be published to other threads.  I don't think that would affect correctness; it just means the rectangle may be calculated more than once.  Not sure if that's a big deal.

Comment (ssoloff) [reply game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java]: FWIW, here's a possible solution that (I think) is thread-safe:

```diff
diff --git a/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java b/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java
index a632f877f..642ce0ba6 100644
--- a/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java
@@ -8,6 +8,10 @@ import java.awt.Polygon;
 import java.awt.Rectangle;
 import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import com.google.common.base.Suppliers;
 
 import games.strategy.engine.data.GameData;
 import games.strategy.engine.data.PlayerID;
@@ -21,7 +25,7 @@ import games.strategy.triplea.ui.mapdata.MapData;
 public class TerritoryNameDrawable implements IDrawable {
   private final String territoryName;
   private final UiContext uiContext;
-  private Rectangle territoryBounds;
+  private final AtomicReference<Supplier<Rectangle>> territoryBoundsSupplier = new AtomicReference<>();
 
   public TerritoryNameDrawable(final String territoryName, final UiContext uiContext) {
     this.territoryName = territoryName;
@@ -81,10 +85,8 @@ public class TerritoryNameDrawable implements IDrawable {
       x = namePlace.get().x;
       y = namePlace.get().y;
     } else {
-      if (territoryBounds == null) {
-        // Cache the bounds since re-computing it is expensive.
-        territoryBounds = getBestTerritoryNameRect(mapData, territory, fm);
-      }
+      territoryBoundsSupplier.compareAndSet(null, Suppliers.memoize(() -> getBestTerritoryNameRect(mapData, territory, fm)));
+      final Rectangle territoryBounds = territoryBoundsSupplier.get().get();
       x = territoryBounds.x + (int) territoryBounds.getWidth() / 2 - fm.stringWidth(territory.getName()) / 2;
       y = territoryBounds.y + (int) territoryBounds.getHeight() / 2 + fm.getAscent() / 2;
     }
```

Although, it's a bit ugly due to the initializer depending on local state.

(Note that the `Supplier` returned by `Suppliers#memoize()` is thread-safe with respect to the cached value.)

**Comment by RoiEXLab** [comment]
@asvitkine You

**Comment by RoiEXLab** [comment]
@asvitkine You're absolutely right about the rectangle. Nevermind then.

File: game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java (hunk 2 above)
Comment (ssoloff) [reply game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java]: Actually, no synchronization may be necessary after all.  Looking up the call stack from `draw()`, it appears the `GameData` lock is being acquired while drawing the image:

https://github.com/triplea-game/triplea/blob/2ff3dd44355d5816287c1bb81a3af44f3cd04ee6/game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java#L575-L583

That introduces a memory barrier and should force the cached value to be published to any other pool thread that happens to access the object in the future.

**TL;DR:** Nevermind. :smile:

**Comment by asvitkine** [comment]
So it there are no further concerns, can this be merged?
